enumerable_methods.rb

```diff
@@ -0,0 +1,219 @@
+      i = 0
+      while i < length
+        yield(self[i])
+        i += 1
+      end
+    elsif self.class == Hash
+      keys = self.keys
+      keys.length.times do |item|
+        key = keys[item]
+        value = self[key]
+        yield(key, value)
+      end
+    end
+    self
+  end
+
+  # my_each_with_index
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    if self.class == Array
+      i = 0
+      while i < length
+        yield(self[i], i)
+        i += 1
+      end
+    elsif self.class == Hash
+      keys = self.keys
+      keys.length.times do |item|
+        key = keys[item]
+        value = self[key]
+        key_value = [key, value]
+        yield(key_value, item)
+      end
+    end
+    self
+  end
+
+  # my_select
+  def my_select
+    return to_enum unless block_given?
+
+    new_arr = []
+    i = 0
+    while i < size
+      new_arr << self[i] if yield(self[i])
+      i += 1
+    end
+    new_arr
+  end
+
+  # my_all
+  # rubocop:disable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+  def my_all?(pattern = nil)
+    arr = self
+    if !block_given? && pattern.nil?
+      arr.my_each do |i|
+        return false unless i
+      end
+    elsif self.class == Array
+      arr.my_each do |i|
+        return false unless yield(self[i])
+      end
+    elsif self.class == Hash
+      keys = self.keys
+      arr.my_each do |i|
+        return false unless yield(keys[i], self[keys[i]])
+      end
+    elsif pattern
+      arr.my_each do |i|
+        return false unless i === pattern
+      end
+    end
+    true
+  end
+  # rubocop:enable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+
+  # my_any
+  # rubocop:disable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+  def my_any?(pattern = nil)
+    arr = self
+    if !block_given? && pattern.nil?
+      arr.my_each do |i|
+        return true if i
+      end
+    elsif self.class == Array
+      arr.my_each do |i|
+        return true if yield(self[i])
+      end
+    elsif self.class == Hash
+      keys = self.keys
+      arr.my_each do |i|
+        return true if yield(keys[i], self[keys[i]])
+      end
+    elsif pattern
+      arr.my_each do |i|
+        return true if i === pattern
+      end
+    end
+    false
+  end
+  # rubocop:enable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+
+  # my_none
+  # rubocop:disable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+  def my_none?(pattern = nil)
+    arr = self
+    if !block_given? && pattern.nil?
+      arr.my_each do |i|
+        return false if i
+      end
+    elsif self.class == Array
+      arr.my_each do |i|
+        return false if yield(self[i])
+      end
+    elsif self.class == Hash
+      keys = self.keys
+      arr.my_each do |i|
+        return false if yield(keys[i], self[keys[i]])
+      end
+    elsif pattern
+      arr.my_each do |i|
+        return false if i === pattern
+      end
+    end
+    true
+  end
```

Comment:
You should use this as a guard case to avoid errors return` self.to_enum unless block_given?`.
There is also this  
 ```
 TypeError:
       no implicit conversion of String into Integer
```